source/common/singleton/threadsafe_singleton.h

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "absl/base/call_once.h"
+
+namespace Envoy {
+
+/**
+ * ThreadSafeSingleton allows easy global cross-thread access to a non-const object.
+ *
+ * This singleton class should be used for singletons which must be globally
+ * accessible but can not be marked as const. All functions in the singleton class
+ * *must* be threadsafe.
```

Comment:
It did occur to me that there is some truth-in-advertising issue with a `ThreadSafeSingleton` that isn't thread safe but forces you to be thread safe. Anyway, potatoes potatahs.